### PR TITLE
Support torch.tensor in script

### PR DESCRIPTION
**Pull request by eellison**
Adding support for torch.tensor in script. 

The input list is typed as t[], because it can be arbitrarily nested. I added a check a compile time check  that the inner type of the list is a bool, float, or int. 

Also adds specialization for Boolean Lists, which already existed at the ivalue level but had not been added to the compiler yet

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -5013,6 +5013,87 @@ def test(op, const, swap_args):
...
+            return torch.tensor(torch.jit.annotate(List[int], []))
+        ''')
+        scope = {}
+        exec(code, globals(), scope)
+        cu = torch.jit.CompilationUnit(code)
+        t1 = cu.func()
+        t2 = scope['func']()
+
+        # torchscript returns int tensor, python returns float tensor
+        self.assertNotEqual(t1.dtype, t2.dtype)
```
Comment (eellison) [inline test/test_jit.py]: Note: python only sees torch.tensor([]), which has a default type of float.  Script uses the annotated type of the list, which is int.

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -2309,7 +2308,7 @@ struct to_ir {
           elem_type = values.at(0)->type();
         }
         for (auto v : values) {
-          if (v->type() != elem_type) {
+          if (!v->type()->isSubtypeOf(elem_type) || !elem_type->isSubtypeOf(v->type()))  {
```
Comment (eellison) [inline torch/csrc/jit/script/compiler.cpp]: Can fix this in a future diff:  int[] == int[] resulted in false

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -2309,7 +2308,7 @@ struct to_ir {
           elem_type = values.at(0)->type();
         }
         for (auto v : values) {
-          if (v->type() != elem_type) {
+          if (*v->type() != *elem_type)  {
```
Comment (wanchaol) [inline torch/csrc/jit/script/compiler.cpp]: what's wrong with here? is it because some `v` might be optional?

File: torch/csrc/jit/register_prim_ops.cpp [hunk 4]
```diff
@@ -928,6 +952,29 @@ Operation listSetItem(const Node* node) {
...
+Operation listSetItem<Shared<BoolList>, bool>(const Node* node) {
+  return [](Stack& stack) {
+    Shared<BoolList> list;
+    int64_t idx;
+    bool value;
+
+    pop(stack, list, idx, value);
+
+    int64_t list_size = list->elements().size();
+    auto normalized_idx = normalizeIndex(idx, list_size);
```
Comment (wanchaol) [inline torch/csrc/jit/register_prim_ops.cpp]: why not just call `getItem` defined above?

File: torch/csrc/jit/register_special_ops.cpp [hunk 5]
```diff
@@ -1,8 +1,11 @@
-#include <torch/csrc/autograd/profiler.h>
-#include <torch/csrc/jit/custom_operator.h>
-#include <torch/csrc/jit/operator.h>
-#include <torch/csrc/api/include/torch/utils.h>
-#include <ATen/ExpandUtils.h>
+#include "torch/csrc/autograd/profiler.h"
+#include "torch/csrc/jit/custom_operator.h"
+#include "torch/csrc/jit/operator.h"
+#include "torch/csrc/api/include/torch/utils.h"
+#include "../ATen/ExpandUtils.h"
```
Comment (wanchaol) [inline torch/csrc/jit/register_special_ops.cpp]: I am not sure if we should use relative includes? this might not consistent with other includes, probably use `aten/src/ATen/ExpandUtils.h`?

**Review by wanchaol — COMMENTED** [review]
LG, Travis and CI need to fix

File: torch/csrc/jit/register_prim_ops.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: getItem returns the value at that index, this is setting the index. I can refactor the getting of the normalized index, though

File: torch/csrc/jit/script/compiler.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: It's working now. Previously it would fail for something like int[] != int[]

File: test/test_jit.py [hunk 6]
```diff
@@ -5001,6 +5001,87 @@ def test(op, const, swap_args):
...
+        with self.assertRaisesRegex(RuntimeError, "Expected sequence of length"):
+            foo()
+
+    def test_torch_tensor_empty_list(self):
+        code = dedent('''
+        def func():
+            return torch.tensor(torch.jit.annotate(List[int], []))
+        ''')
+        scope = {}
+        exec(code, globals(), scope)
```
Comment (driazati) [inline test/test_jit.py]: Does this have to use `exec`? It looks like it would be fine as a normal @torch.jit.script function

File: torch/csrc/jit/register_prim_ops.cpp [hunk 7]
```diff
@@ -751,6 +750,15 @@ typename TList::element_type::ElemType& getItem(TList& list, int64_t idx) {
   return list->elements()[normalized_idx];
 }
 
+bool getItem(const std::vector<bool>& list, int64_t idx) {
+  const int64_t list_size = list.size();
+  const int64_t normalized_idx = normalizeIndex(idx, list_size);
+  if (normalized_idx < 0 || normalized_idx >= list_size) {
```
Comment (driazati) [inline torch/csrc/jit/register_prim_ops.cpp]: `std::vector::at()` does the range check for you afaik

File: torch/csrc/jit/register_special_ops.cpp [hunk 8]
```diff
@@ -11,6 +14,126 @@ namespace torch {
...
+  checkSequenceSize(n, dim, seq_size);
+  for (int64_t i = 0; i < n; i++) {
+    *(DTYPE*)data = obj[i];
+    data += strides[dim] * elementSize;
+  }
+}
+
+// bool vector needs to be cast to uint8_t
+template<>
+void storeLastDimension<bool>(char* data, const std::vector<int64_t>& sizes, const c10::ArrayRef<long>& strides, int64_t dim,
```
Comment (driazati) [inline torch/csrc/jit/register_special_ops.cpp]: Why is this specialization necessary? Shouldn't it be fine since `char` and `uint8_t` are defined to both be 1 byte?

**Review by driazati — APPROVED** [review]
Looks good, comments are all surface level stuff

File: torch/csrc/jit/register_special_ops.cpp (hunk 8 above)
Comment (eellison) [reply torch/csrc/jit/register_special_ops.cpp]: This line:
```
*(uint8_t*)data = static_cast<uint8_t>(obj[i]);
```

There needs to be the cast from bool to uint8. From what I googled / saw on stack overflow, static_cast on the same type isn't guaranteed to be a noop, plus having extra casts when they're not needed seems like misleading code.

File: test/test_jit.py [hunk 9]
```diff
@@ -5001,6 +5001,106 @@ def test(op, const, swap_args):
...
         self._test_tensor_number_math()
 
+    def test_torch_tensor_bad_input(self):
+        with self.assertRaisesRegex(RuntimeError, "Input list to torch.tensor must be of ints, floats, "
+                                    "or bools, got None"):
+            @torch.jit.script
+            def test():
+                return torch.tensor([None])
+
+        with self.assertRaisesRegex(RuntimeError, "Note: empty lists are constructed as Tensor"):
```
Comment (zou3519) [inline test/test_jit.py]: What does this error message mean?

File: test/test_jit.py (hunk 6 above)
Comment (zou3519) [reply test/test_jit.py]: You don't need to exec this, the `exec` is useful when one needs to format a string to produce a test case but that doesn't occur here

File: test/test_jit.py (hunk 1 above)
Comment (zou3519) [reply test/test_jit.py]: We shouldn't have this discrepancy? Or at least, there should be a warning here

Comment (eellison) [reply test/test_jit.py]: The first time a user inputs torch.tensor([]) will result in exception saying they have to annotate what type of list they want. After the user then annotates it as type int, I think it would be strange to return a float tensor. What warning would you suggest ?

File: test/test_jit.py (hunk 9 above)
Comment (eellison) [reply test/test_jit.py]: It's the same error message as in the compiler:
```
empty lists are constructed as Tensor[]; \n"
               << "if you want an empty list of a different type, \n"
               << "use `torch.jit.annotate(List[T], [])`, \n"
               << "where `T` is the type of elements in the list)";
```
We type empty lists as tensor lists by default. So if a user tries torch.tensor([]) the more informative error message is printed, in addition to `Input list to torch.tensor must be of ints, floats, or bools, go" << type->str().

File: test/test_jit.py (hunk 1 above)
Comment (zou3519) [reply test/test_jit.py]: I think a general rule of thumb should be that code in script should run the same as code without script. If the user annotated the list as int without script, normal pytorch would still return an empty float tensor -- because of that discrepancy, we should warn when `torch.tensor([])` returns something that is not a float tensor (or the default tensor type)

File: test/test_jit.py (hunk 9 above)
Comment (zou3519) [reply test/test_jit.py]: Got it, thanks for the explanation, I forgot that you only need to pass part of the error message to assertRaisesRegex

File: torch/csrc/jit/register_special_ops.cpp [hunk 10]
```diff
@@ -1,8 +1,11 @@
...
-#include <torch/csrc/api/include/torch/utils.h>
-#include <ATen/ExpandUtils.h>
+#include "torch/csrc/autograd/profiler.h"
+#include "torch/csrc/jit/custom_operator.h"
+#include "torch/csrc/jit/operator.h"
+#include "torch/csrc/api/include/torch/utils.h"
+#include "aten/src/ATen/ExpandUtils.h"
+#include "c10/core/ScalarType.h"
+#include "aten/src/ATen/InitialTensorOptions.h"
+#include "torch/csrc/jit/script/error_report.h"
```
Comment (zou3519) [inline torch/csrc/jit/register_special_ops.cpp]: We're moving away from "" includes and instead moving to <> includes I think

File: test/test_jit.py (hunk 1 above)
Comment (eellison) [reply test/test_jit.py]: Added warning - let me know what you think

Comment (zou3519) [reply test/test_jit.py]: Looks better now! One suggestion:

`torch.empty([])` can also return a `DoubleTensor` (as opposed to float tensor) if one changed the default dtype: https://pytorch.org/docs/master/torch.html#torch.set_default_dtype

Maybe you should warn that the "list will create a tensor of default floating point type (currently <>) in python but a tensor of type <> in torchscript"
